Add screencapture

Ealdwulf opened the pull request:
Thought I'd make a PR before this gets to large. The command line options etc may still change, and there are known issues.

This PR:
*  Adds a kano-screencapture binary to kano-screenshot which feeds the output of `dispmanx_spanshot` to the h264 encoder.
* `src/ilclient/*` is taken directly from `libraspberrypi-doc`
* src/kano-screencapture.cpp is taken from `encode.cpp` in `libraspberrypi-doc` and then modified.
* all dispmanx code is factored out of kano-screenshot and used in both.
* As a side effect of adding support for VC_IMAGE_BGR888 which seems to be the best option to feed to the encoder, we end up with a new -t BGR888 option in kano-screenshot, but libpng doesn't like it so it's not functional (just produces an error message).
* Currently it can be stopped early using ctrl-c, but I'm not sure that's strictly reliable (not sure all the libraries retry system calls properly if they get EAGAIN) so I may change it later to input from stdin or something.
* The output is raw h264. May change later to mp4.

Ealdwulf committed:
Separate out screen grabbing function. The idea is to make it into
a library.

Ealdwulf committed:
Add copy of ilclient library from libraspberrypi-doc.

Ealdwulf committed:
get hello_encode compileing (but named kano-screencapture).

Ealdwulf committed:
First working version of kano-screencapture.
Fixed output size, probably doesn't work yet on different screen sizes.

Ealdwulf committed:
Use BGR888, which eliminates the loss of precision from using RBG565.
The encoder doesn't support RGB888, but fortunately both sides
support BGR888.

Ealdwulf committed:
Include kano-screencapture in debian package.
Add command line arguments for size, rate, number of frames.
Add stopmotion mode with sigHUP.
Add early termination with sigINT.

Ealdwulf committed:
Rename encode.cpp to kano-screencapture.cpp

Ealdwulf committed:
Add changelog entry.

Ealdwulf committed:
Remove stopmotion mode.

tombettany commented on src/dispmanx_grabber.cpp:
```diff
@@ -0,0 +1,194 @@
+    { "RGBA16", VC_IMAGE_RGBA16, 2 },
+    { "RGBA32", VC_IMAGE_RGBA32, 4 }
+};
+static size_t imageEntries = sizeof(imageInfo)/sizeof(imageInfo[0]);
```
`const`?

Ealdwulf replied:
Yes.

tombettany commented on src/dispmanx_grabber.cpp:
```diff
@@ -0,0 +1,194 @@
+	  {
+            double numerator = height * cfg.requestedWidth;
+            double denominator = width;
+            height = (int)ceil(numerator / denominator);
```
This bit is super confusing. Presumably what we are trying to do is to work out if we scaled down the width based on the min operation above and, if so, scale the height by the same factor. Maybe it would be clearer to write something along the lines of:

```cpp
double scale = static_cast<double>(cfg.requestedWidth) / static_cast<double>(width);
height = static_cast<int>(ceil(height * scale))
```

(also old C-style casts in a C++ program).

The other thing that makes this confusing is that when read in conjunction with the `if` clause just below, it isn't obvious that because we are in this current `if` statement that we will miss the subsequent one so I was very confused about whether there would be some contamination of the `width`, `height`, `cfg.requestedWidth` and `cfg.requestedHeight` variables. I don't know if there is anything that can be done to improve this but that's how it reads.

Ealdwulf replied:
Yeah, that's all pretty confusing, I'll have a think about it.

tombettany commented on src/dispmanx_grabber.cpp:
```diff
@@ -0,0 +1,194 @@
+	  {
+	    double numerator = width * cfg.requestedHeight;
+            double denominator = height;
+            width = (int)ceil(numerator / denominator);
```
Obviously, same here.

tombettany commented on src/dispmanx_grabber.cpp:
```diff
@@ -0,0 +1,194 @@
+    st->resourceHandle = vc_dispmanx_resource_create(st->frameInfo.imageType,
+                                                     width,
+                                                 height,
+                                                 &st->vcImagePtr);
```
Indentation is all over the place here.

Ealdwulf replied:
Will fix.

tombettany commented on src/dispmanx_grabber.cpp:
```diff
@@ -0,0 +1,194 @@
+    // the only implemented option at this API level (see rotate_image_180 function).
+    int result = vc_dispmanx_snapshot(st->displayHandle,
+                                      st->resourceHandle,
+                                  DISPMANX_NO_ROTATE);
```
Indentation

tombettany reviewed (approved):
Only minor comments. Oh and copyright headers everywhere :p

radujipa reviewed (approved):
Make sure to use our copyright header where appropriate on any new files added :shipit:

Ealdwulf commented on src/Makefile:
```diff
@@ -10,17 +10,22 @@
-CFLAGS+=-Wall -std=c++0x $(shell libpng-config --cflags)
-LDFLAGS+=-L/opt/vc/lib/ -lbcm_host $(shell libpng-config --ldflags) -lX11
-INCLUDES+=-I/opt/vc/include/ -I/opt/vc/include/interface/vcos/pthreads -I/opt/vc/include/interface/vmcs_host/linux
+CFLAGS+=-Wall -std=c++0x $(shell libpng-config --cflags) -DSTANDALONE -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -DTARGET_POSIX -D_LINUX -fPIC -DPIC -D_REENTRANT -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -U_FORTIFY_SOURCE  -DHAVE_LIBOPENMAX=2 -DOMX -DOMX_SKIP64BIT -ftree-vectorize -pipe -DUSE_EXTERNAL_OMX -DHAVE_LIBBCM_HOST -DUSE_EXTERNAL_LIBBCM_HOST -DUSE_VCHIQ_ARM -Wno-psabi -Wwrite-stri ...
```
We might want to, but it's not trivial as they share an object. I'll keep an eye on it; if kano-screencapture grows lots more stuff then I'll move dispmanx into  a separate directory and do that.

Ealdwulf commented on src/dispmanx_grabber.cpp:
```diff
@@ -0,0 +1,194 @@
+
+#define kprintf if(st->cfg.verbose && st->cfg.logPrintf) st->cfg.logPrintf
+
+static DispManxImageInfo imageInfo[] =
```
👍

Ealdwulf commented:
@radujipa Good spot. Since ilclient/* are unmodified, I'll not add the copyright header to those - we can add it if we change them.

Ealdwulf committed:
Add copyright notices.
Make requested height/width calc more comprehensible.
Add consts.
Fix formatting.

Ealdwulf commented:
@tombettany Should be more comprehensible now...
